gateway/src/apicast/policy/keycloak_role_check/keycloak_role_check.lua

```diff
@@ -0,0 +1,195 @@
+--- Keycloak Role Check Policy
+-- This policy verifies the realm roles and the client roles in the JWT.
+--
+--- The realm roles are specified when you want to add role check to the every client's resources or 3scale itself.
+-- https://www.keycloak.org/docs/4.0/server_admin/index.html#realm-roles
+--
+-- When you specify the realm roles in Keycloak, the JWT includes them as follows:
+--
+-- {
+--   "realm_access": {
+--     "roles": [
+--       "<realm_role_A>", "<realm_role_B>"
+--     ]
+--   }
+-- }
+--
+-- And you need to specify the "realm_roles" in this policy as follows:
+--
+-- "realm_roles": [
+--   { "name": "<realm_role_A>" }, { "name": "<realm_role_B>" }
+-- ]
+--
+--- The client roles are specified when you want to add role check to the particular client's resources.
+-- https://www.keycloak.org/docs/4.0/server_admin/index.html#client-roles
+--
+-- When you specify the client roles in Keycloak, the JWT includes them as follows:
+--
+-- {
+--   "resource_access": {
+--     "<client_A>": {
+--       "roles": [
+--         "<client_role_A>", "<client_role_B>"
+--       ]
+--     },
+--     "<client_B>": {
+--       "roles": [
+--         "<client_role_A>", "<client_role_B>"
+--       ]
+--     }
+--   }
+-- }
+--
+-- And you need to specify the "client_roles" in this policy as follows:
+--
+-- "client_roles": [
+--   { "name": "<client_role_A>", "client": "<client_A>" },
+--   { "name": "<client_role_B>", "client": "<client_A>" },
+--   { "name": "<client_role_A>", "client": "<client_B>" },
+--   { "name": "<client_role_B>", "client": "<client_B>" }
+-- ]
+
+local policy = require('apicast.policy')
+local _M = policy.new('Keycloak Role Check Policy')
+
+local ipairs = ipairs
+local MappingRule = require('apicast.mapping_rule')
+local TemplateString = require('apicast.template_string')
+local default_type = 'plain'
+
+local new = _M.new
+
+local function build_templates(scopes)
+  for _, scope in ipairs(scopes) do
+
+    if scope.realm_roles then
+      for _, realm_role in ipairs(scope.realm_roles) do
+        realm_role.template_string = TemplateString.new(
+          realm_role.name, realm_role.name_type or default_type)
+      end
+    end
+
+    if scope.client_roles then
+      for _, client_role in ipairs(scope.client_roles) do
+        client_role.name_template_string = TemplateString.new(
+          client_role.name, client_role.name_type or default_type)
+
+        client_role.client_template_string = TemplateString.new(
+          client_role.client, client_role.client_type or default_type)
+      end
+    end
+
+    scope.resource_template_string = TemplateString.new(
+      scope.resource, scope.resource_type or default_type)
+
+  end
+end
+
+function _M.new(config)
+  local self = new()
+  self.config = config or {}
```

Comment:
I don't think this is right.
`self.config` is assigned here, but it's never used, and then the code assumes that `config.type` and `config.scopes` exist.

Reply:
Fixed.